### DISCUSSION TITLE
discrete-time systems

### DISCUSSION THREAD

**Discussion by artpelling** [pymor/pymor#1499] [Ideas]
Hello everyone, I am starting to work on a rudimentary implementation of discrete-time LTI systems #1500 .
Right off the bat, there are some design choices to be made regarding the `cont_time` attribute.

The initial idea was to rename `cont_time` to `dt`. The Python Control Systems Library does this such that `dt` can be `None` or `0` for continuous-time systems or a number or `True` for discrete-time systems, where `True` stands for discrete-time with unspecific sampling time.
While I like the inclusion of `True`, I dislike the ambiguity of `None` and `0` for continuous time, because checking for that (when raising `NotImplementedError` or switching of computational methods) will be error-prone. The way I see it, there are a couple of ways to go about it.

1. remove `0` and just have `dt=None` for continuous-time systems so that it can easily be checked.
2. keep the `cont_time` attribute and have an additional `dt` parameter that can be `0`,`None`, a number or `True`

A second issue is whether or not physical units (especially time) should be included as attributes such that they can be used in the plotting routines.

**Comment by ftalbrecht** [comment]
Hi,

we should probably discuss this in line with changes to the yet-to-envisioned `TimedependentModel`. There is currently no well-defined way to obtain the length of a state trajectory returned by `solve` before calling `solve` (while information about its dimension can be obtained from `solution_space`, if that space is fixed for all inputs). One usually has to write something like
```python
len_trajectory = m.time_stepper.num_values or m.time_stepper.nt + 1
```
since currently all `InstationaryModel`s have a time stepper.
Maybe, we want to have something simpler in the interface, similar to what is proposed above.

**Reply by artpelling** [reply]
In `dmd.py` a `cont_time_dt` attribute is used. I suggest that this should be adjusted as well.

**Comment by pmli** [comment]
It seems to me that [python-control](https://github.com/python-control/python-control/blob/c2f6f8ab94bbc8b5ef1deb33c3d2df39e00d22bf/control/lti.py#L35-L38) uses `dt=None` to denote either a continuous-time or a discrete-time system. On the other hand, [`scipy.signal`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.StateSpace.html) uses `None` for continuous-time systems and a positive number for discrete-time systems (and also [`True`](https://github.com/scipy/scipy/blob/47bb6febaa10658c72962b9615d5d5aa2513fa3a/scipy/signal/ltisys.py#L1723-L1725), it seems).

I kind of like the idea of using only floats for `dt`, i.e., `dt=0` for continuous-time systems and `dt>0` for discrete-time systems. Do you have a use case for unspecified sampling time? What would then `freq_resp` do?

**Reply by ftalbrecht** [reply]
I'm not 100% sure (maybe @sdrave knows better), but unspecified sampling time might correspond to a `TimedependentModel` with a time-stepper where the time step size depends on the input parameter (i.e., a parameterized advection term). Thus, its discrete in time, but the length of the trajectory might change in each call.

**Reply by pmli** [reply]
I'm not sure that we are talking about the same thing. In systems theory, continuous-time and discrete-time systems are treated rather differently, e.g., the frequency response of a continuous-time system is given by the values of its transfer function over the imaginary axis, while for discrete-time it is over the unit circle (this is where the sampling time enters, because frequency is proportional to `1/dt`). Also, balanced truncation and H2-optimal MOR are somewhat different for these two cases (interestingly, for discrete-time systems, they do not need the value of `dt`).

BTW, discrete-time systems do not necessarily come from a time discretization of a continuous-time system.

For discrete-time systems, it's a bit strange to talk about time-steppers, since there is just a difference equation that needs to be propagated.

We should definitely discuss how all this should be treated when connecting `InstationaryModel` and `LTIModel`, but maybe we can do this once the implementation for `LTIModels` is done?

**Reply by artpelling** [reply]
I like an only-float `dt` attribute very well, too.

Regarding `True`: (I have not thought this through completely yet but) I was thinking towards a discrete-time system with normalized frequency. But this would become relevant only when plotting the signal, so this functionality is probably best implemented there.

I will have to dissect the topic of handling the frequencies more. At the moment `freq_resp` allows for arbitrary frequencies as inputs. The user inputs would have to be restricted to valid frequencies or a suitable mapping has to be made (in a way this is already done by silently multiplying by `1j` in the case of continuous-time systems).
~~Maybe the slimmest version would be to use integers as frequencies and let `dt` come into play when plotting?~~

**Reply by artpelling** [reply]
I have implemented my proposed changes to `freq_resp` and the plotting routines. For now, I have left out `dt=True`. Integer values for `dt` are cast to float.

I suggest to include a user warning in `freq_resp` when the angular input  frequencies exceed Nyquist. I have found #1468, is there somewhere I can read up on the coding style regarding user warnings?

**Comment by pmli** [comment]
> A second issue is whether or not physical units (especially time) should be included as attributes such that they can be used in the plotting routines.

I would be just for assuming that `dt` is given in seconds. It seems that this is done in `scipy.signal` and `python-control`. Do you have a use case in mind for which it would be good to have physical units?

**Reply by artpelling** [reply]
Absolutely, I was thinking mainly about the physical units of the input and output quantities in order to get physically correct y-axis scalings. I think Matlab has implemented this. On second thoughts, this is probably also relevant for non LTI systems and would bloat the code for a small "nice-to-have"

**Reply by pmli** [reply]
Hm, I'm not familiar with this. I'm also not sure what this would mean for pyMOR (e.g., does this affect solvers or only plotting). But, this sounds like a separate issue from support for discrete-time systems, so adding a different issue or discussion would be good if you are interested in this.

**Reply by drittelhacker** [reply]
I could imagine that this also affects, e.g., BT Reductors, since a change of units (say from meters to milimeters) may be used for scaling purposes, which in turn influences the error (bound) and Gramian computation, but can lead to better reduction results or smaller models...

**Reply by artpelling** [reply]
It is just a thing that came to mind when tinkering with the plotting routines. I have encountered this feature in the Matlab Control Systems Library in which they use strings to represent state, time, input and output units.

As @drittelhacker suggests, this could make it easier (e.g. in engineering applications) to use physically meaningful reduction error bounds and create more easily interpretable plots without headaches.

However, IMO this would only be worth the effort, if SI units are be implemented symbolically such that interconnected systems would automatically have the correct units (maybe with the siunits package?). I will keep this in mind and maybe create a feature request in the future if it stays relevant!

**Reply by pmli** [reply]
I never used physical units, but I heard that there are a bunch of packages that help with those (I think a popular one is [`pint`](https://github.com/hgrecco/pint)).

I'm still not sure how units would be used (assigning units to states sounds tricky for large-scale systems), so a concrete proposal would be great.
